shopfloor/services/zone_picking.py

```diff
@@ -0,0 +1,887 @@
+    ):
+        domain = [
+            ("location_id", "child_of", location.id),
+            ("qty_done", "=", 0),
+            ("state", "in", ("assigned", "partially_available")),
+        ]
+        if picking_type:
+            domain += [("picking_id.picking_type_id", "=", picking_type.id)]
+        if package:
+            domain += [("package_id", "=", package.id)]
+        if product:
+            domain += [("product_id", "=", product.id)]
+        if lot:
+            domain += [("lot_id", "=", lot.id)]
+        return domain
+
+    def _find_location_move_lines(
+        self, location, picking_type=None, package=None, product=None, lot=None
+    ):
+        """Find lines that potentially are to move in the location"""
+        return self.env["stock.move.line"].search(
+            self._find_location_move_lines_domain(
+                location, picking_type, package, product, lot
+            )
+        )
+
+    def scan_location(self, barcode):
+        """Scan the zone location where the picking should occur
+
+        The location must be a sub-location of one of the picking types'
+        default source locations of the menu.
+
+        Transitions:
+        * start: invalid barcode
+        * select_picking_type: the location is valid, user has to choose a picking type
+        """
+        search = self.actions_for("search")
+        zone_location = search.location_from_scan(barcode)
+        if not zone_location:
+            return self._response_for_start(message=self.msg_store.no_location_found())
+        if not zone_location.is_sublocation_of(
+            self.work.menu.picking_type_ids.mapped("default_location_src_id")
+        ):
+            return self._response_for_start(
+                message=self.msg_store.location_not_allowed()
+            )
+        move_lines = self._find_location_move_lines(zone_location)
+        if not move_lines:
+            return self._response_for_start(
+                message=self.msg_store.no_lines_to_process()
+            )
+        picking_types = move_lines.picking_id.picking_type_id
+        return self._response_for_select_picking_type(zone_location, picking_types)
+
+    def list_move_lines(self, zone_location_id, picking_type_id, order="priority"):
+        """List all move lines to pick, sorted
+
+        Transitions:
+        * select_line: show the list of move lines
+        """
+        zone_location = self.env["stock.location"].browse(zone_location_id)
+        if not zone_location.exists():
+            return self._response_for_start(message=self.msg_store.record_not_found())
+        picking_type = self.env["stock.picking.type"].browse(picking_type_id)
+        if not picking_type.exists():
+            return self._response_for_start(message=self.msg_store.record_not_found())
+        move_lines = self._find_location_move_lines(zone_location, picking_type)
+        # TODO 'order' handling needs to be refactored
```

Comment:
@sebalix it might be probably better to not do this on server side. I can easily do it on client side. 
Do you have any specific requirements to keep this server side?

Reply:
Checked on Slack, we preserve the order server-side.